station/waterfall.py

```diff
@@ -0,0 +1,156 @@
+# This file is heavily based on waterfall.py from satnogs-client.
+
+import matplotlib
+import numpy as np
+import sys
+
+matplotlib.use('Agg')
+
+import matplotlib.pyplot as plt
+
+OFFSET_IN_STDS = -2.0
+SCALE_IN_STDS = 8.0
+
+class Waterfall():
+    """
+    Parse waterfall data file
+
+    :param datafile_path: Path to data file
+    :type datafile_path: str_array
+    """
+    def __init__(self, datafile_path, logger=None):
+        """
+        Class constructor
+        """
+        self.logger = logger or sys.stdout
+        self.data = self._get_waterfall(datafile_path)
+
+    def plot(self, figure_path, vmin=None, vmax=None):
+        """
+        Plot waterfall into a figure
+
+        :param figure_path: Path of figure file to save
+        :type figure_path: str
+        :param value_range: Minimum and maximum value range
+        :type value_range: tuple
+        """
+        tmin = np.min(self.data['data']['tabs'] / 1000000.0)
+        tmax = np.max(self.data['data']['tabs'] / 1000000.0)
+        fmin = np.min(self.data['freq'] / 1000.0)
+        fmax = np.max(self.data['freq'] / 1000.0)
+
+        self.logger.write(f"vmin=%{vmin} vmax={vmax} tmin={tmin} tmax={tmax} fmin={fmin} fmax={fmax}")
+        if vmin is None or vmax is None:
+            vmin = -100
+            vmax = -50
+            c_idx = self.data['data']['spec'] > -200.0
+            if np.sum(c_idx) > 100:
+                data_mean = np.mean(self.data['data']['spec'][c_idx])
+                data_std = np.std(self.data['data']['spec'][c_idx])
+                vmin = data_mean - 2.0 * data_std
+                vmax = data_mean + 4.0 * data_std
+        plt.figure(figsize=(10, 20))
+        plt.imshow(self.data['data']['spec'],
+                   origin='lower',
+                   aspect='auto',
+                   interpolation='None',
+                   extent=[fmin, fmax, tmin, tmax],
+                   vmin=vmin,
+                   vmax=vmax,
+                   cmap='Greens') # 'viridis', 'plasma', 'inferno', 'magma', 'cividis'
+                                  # also, see https://matplotlib.org/stable/tutorials/colors/colormaps.html
+        plt.xlabel('Frequency (kHz)')
+        plt.ylabel('Time (seconds)')
+        fig = plt.colorbar(aspect=50)
+        fig.set_label('Power (dB)')
+        plt.savefig(figure_path, bbox_inches='tight')
+        plt.close()
+
+    def _read_waterfall(self, datafile_path):
+        """
+        Read waterfall data file
+
+        :param datafile_path: Path to data file
+        :type datafile_path: str
+        :return: Waterfall data
+        :rtype: dict
+        """
+
+        datafile = open(datafile_path, mode='rb')
```

Comment:
Always use ```open``` with ```with``` context manager. In this function ```datafile``` isn't closed when an exception is thrown.

Reply:
Updated as suggested.